### PR TITLE
feat: Add support for storj.io

### PR DESCRIPTION
**Pull request by damien-abos**
Allow to use storj.io with End-to-End encryption and a distributed file storage
See https://www.storj.io/how-it-works

### CODE REVIEW

File: db.go [hunk 1]
```diff
@@ -1464,7 +1464,7 @@ func (db *DB) CalcRestoreTarget(ctx context.Context, opt RestoreOptions) (*Repli
 		}
 
 		// Use the latest replica if we have multiple candidates.
-		if !updatedAt.After(target.updatedAt) {
+		if generation == target.generation && !updatedAt.After(target.updatedAt) {
```
Comment (hifi) [inline db.go:1467]: Why was this change needed?

**Review by hifi — COMMENTED** [review]
The author has stated before that if a provider has S3 compatibility no new integrations will be added to Litestream and storj.io seems to.

File: db.go (hunk 1 above)
Comment (damien-abos) [reply db.go:1467]: I had an issue when restoring a database with a single generation and a target.updateAt set at 0, then the condition was true and at the end there was no replica available.

**Comment by hifi** [comment]
This PR might suffer from #535 which should be checked before merging as the SDK doesn't seem to say anything about object order.

**Comment by hifi** [comment]
Okay #538 fixes Storj through their S3 gateway and probably this PR as well.

I suppose there's still an argument to use their native SDK for client side encryption (even though Litestream currently supports age encryption as well) and direct access to the network rather than through a gateway.

**Comment by benbjohnson** [comment]
@damien-abos I'm going close out this PR as it adds in additional client dependencies which will inflate the binary size. It sounds like there's an S3 gateway as well which would work with Litestream.

File: db.go (hunk 1 above)
Comment (hifi) [reply db.go:1467]: Even though the PR itself was rejected I'm interested in understanding this issue if it's reproducible with the S3 gateway of Sorj?

Comment (hifi) [reply db.go:1467]: Looking into the code I believe the problem was that your replica implementation returned a generation without updatedAt set to the latest snapshot/WAL file which would've been wrong as otherwise it would always be >0.
